### ISSUE TITLE
defaultMutations dont see state variable and it's not clear how to implement it

### ISSUE THREAD

**Issue by artfabrique** [mesqueeb/vuex-easy-access#36]
Hi! I'm a bit confused.
When i add `...defaultMutations(state)` to `mutations` object it gives me a reference error.
Same happens whei i try this not in the module but in the store.js itsefl (where all the vuex stuff starts)

```
// file /store/modules/userState.js

import {defaultMutations} from 'vuex-easy-access'

const userState = {
    firestorePath: 's1-users/nomodel',
    firestoreRefType: 'doc', // or 'doc'
    moduleName: 'userState',
    statePropName: 'data',
    namespaced: true, // automatically added

    // this object is your store module (will be added as '/myModule')
    // you can also add state/getters/mutations/actions
    state: {
        user: "",
        referrer: "",
        created_at: "",
        created_by: "",
        id: "",
        last_login: "",
        last_session_id: "",
        
        //device
        device_model: "",
        device_type: "",
        os_version: "",
        os: "",

        //survey
        
    },
    getters: {},
    mutations: {
        ...defaultMutations(state)
    },
    actions: {},
}
export default userState
```

Its not clear from any of the documentation where how to use defaultMutations and where to inject it.

Here is my store.js
```
// file /store/store.js

import Vue from "vue";
import Vuex from "vuex";
import VuexEasyFirestore from "vuex-easy-firestore";
import createEasyAccess from "vuex-easy-access";
import { defaultMutations } from "vuex-easy-access";

Vue.use(Vuex);

import { Firebase, initFirebase } from "@/firebase.js";
import userState from "./modules/userState.js";
import state from "vuex-easy-firestore/types/module/state";

const easyAccess = createEasyAccess({
  vuexEasyFirestore: true
});

const easyFirestore = VuexEasyFirestore([userState], {
  logging: true,
  FirebaseDependency: Firebase
});

const AppLifecycleState = {
  LOADING: "loading",
  LOGGING_IN: "loggingIn",
  LOADING_USER_DATA: "loadingUserData",
  INTRO: "intro",
  SENDING_RESPONSE: "sendingResponse",
  LOADING_NEXT_CARD: "loadingNextCard",
  TOUR_FINISHED: "tourFinished",
  ALL_FINISHED: "allFinished",
  NICKNAME_AND_SOCIAL_LOGIN: "nicknameAndSocialLogin",
  SENDING_NICK_AND_SOCIAL: "sendingNickAndSocial",

  ERROR_COMMON: "errorCommon",

  check: function(value) {
    if (value == undefined || value == null) {
      throw new Error("AppLifecycleState: passed Enum is " + value + "!");
    }
    for (const key in this) {
      if (this.hasOwnProperty(key)) {
        const element = this[key];
        if (element == value && typeof element == "string") {
          return true;
        }
      }
    }
    return false;
  }
};

const store = new Vuex.Store({
  strict: true,
  state: {
    prevAppState: undefined,
    currentAppState: AppLifecycleState.LOADING,
    user: null,
    cards: null
  },
  mutations: {
    updateAppState(state, newAppState) {
      if (!AppLifecycleState.check(newAppState))
        throw new Error(
          "Can not set app state to " + newAppState + ": no such state!"
        );
      if (store.get("currentAppState") != newAppState)
        store.set("prevAppState", store.get("currentAppState"));
      store.set("currentAppState", newAppState);
    },
    loginOrRegisterUser() {},

    ...defaultMutations(state)
  },
  getters: {
    user: state => state.user
  },
  plugins: [easyFirestore, easyAccess]
});

initFirebase().catch(error => {
  console.error("FB ERROR!" + error);
});

export { store, AppLifecycleState };
```

**Comment by artfabrique** [comment]
It seems that I've managed it to work in the module with this code
I was not clear that you have to pass **module store** to the `defaultMutations()` and not the **app store**.
In the documentation, it is said that I have to pass a state, not the whole store.

But it is still don't clear how to generate auto mutations for the root store object.
Thy just dont get generated :(

```js
import {defaultMutations} from 'vuex-easy-access'

const userState = {
    firestorePath: 's1-users/nomodel',
    firestoreRefType: 'doc', // or 'doc'
    moduleName: 'userState',
    statePropName: 'data',
    namespaced: true, // automatically added

    // this object is your store module (will be added as '/myModule')
    // you can also add state/getters/mutations/actions
    state: {
        user: "",
        referrer: "",
        created_at: "",
        created_by: "",
        id: "",
        last_login: "",
        last_session_id: "",
        
        //device
        device_model: "",
        device_type: "",
        os_version: "",
        os: "",

        //survey
        
    },
    getters: {},
    mutations: {
        ...defaultMutations(userState)
    },
    actions: {},
    //plugins: [easyFirestore]
}
export default userState
```

**Comment by mesqueeb** [comment]
Hi @artfabrique 
Sorry for my late response.
You're right that it's not that clear from the documentation, I will improve on that.

Basically you just need to define the state outside of the module and pass that same object to both the state and the mutations. : )

See how it's done here:
```js
const state = {
  user: "",
  // ...
}

const userState = {
  state: state, // pass state to state
  mutations: {
    ...defaultMutations(state) // pass as param to defaultMutations
  },
}
```
Sorry for being unclear! 😅 
I hope this helps!!

**Comment by artfabrique** [comment]
@mesqueeb  Yeah! I've just figured it out!
Thanks a lot, great package. I'll try to report any uncertainty that I will find with a "newbie eye". I'm also heavily using (try to use) your other package (easy firebase)

**Comment by mesqueeb** [comment]
No problem! And thanks for helping out : )
I'll close this issue when I update the docs!

PS, some advice: **I highly suggest keeping usage of this library to a bare minimum.**
(over-using it might lead to parts of your app to become more difficult to reason about, because it's less clear how changes are made)

If you mainly use vuex-easy-firestore you will receive actions like
- `dispatch('myModule/set')`
- `dispatch('myModule/patch')`
- `dispatch('myModule/insert')`
- `dispatch('myModule/delete')`

and vuex-easy-access adds
- default mutations for state props
- a submodule called `/set/`

It's this submodule that is often confusing when combining with vuex-easy-firestore.

See your specific case:

```js
const userState = {
    firestorePath: 's1-users/nomodel',
    firestoreRefType: 'doc', // or 'doc'
    moduleName: 'userState',
    statePropName: 'data',
    state: {
        user: '',
    }
}
```
If this module is made usable with both libraries you will be able to do:
```js
dispatch('userState/set/user', newUser) // easy-access
// or
dispatch('userState/set', {user: newUser}) // easy-firestore
```
In this case "easy-access" will detect "easy-firestore" and call that action of "easy-firestore" instead of _just_ mutating the state.

However, if you start using both these methods at random throughout your app you end up having **inconsistency**, which is something that easily makes your app harder to read.

That's why I'd actually suggest what I personally do:
- always use the vuex-easy-firestore method of modifying your state.
- only use vuex-easy-access for the (auto-generated) default mutations
  - eg. `commit('userState/user', newUser)`
  - Using an (auto-generated) default mutation like this allows you to change the state without triggering a sync to firestore.

So basically, if you decide _not_ to use the `/set/` sub-module of easy-access, you can just **not pass vuex-easy-access to the store plugins**, and only import the `defaultMutations` function from this library.

Do you understand? 😅
Just some advice though....


<sub>--<br>Vuex Easy Firestore was made with ♥ by Luca Ban.<br>If this library helped you in any way you can support me by buying me <a href="https://www.buymeacoffee.com/mesqueeb">a cup of coffee</a>.</sub>
